src/providers/oracle/qgsoraclefeatureiterator.cpp

```diff
@@ -528,6 +528,31 @@ bool QgsOracleFeatureIterator::openQuery( const QString &whereClause, const QVar
   return true;
 }
 
+bool QgsOracleFeatureIterator::execQuery( const QString &query, const QVariantList &args, int retryCount )
+{
+  if ( !QgsOracleProvider::exec( mQry, query, args ) )
+  {
+    if ( retryCount != 0 )
+    {
+      // If the connection has been closed
+      // Try again N times in case of timeout
+      // ORA-12170: TNS:Connect timeout occurred
+      if ( mQry.lastError().number() == 12170 )
+      {
+        // remove connection
+        mConnection->disconnect();
+        // get the new connection
+        mConnection = QgsOracleConn::connectDb( mUri );
```

Comment:
Doesn't build either:

```suggestion
        mConnection = QgsOracleConn::connectDb( mSource->mUri );
```

Did you actually try any of this?  `connectDb()` will simply return the existing (defunct) connection from the pool unless all references are cleared upfront.

Reply:
you are right, I need to close and repon the database connection, as explain here https://stackoverflow.com/questions/3615422/mysql-server-has-gone-away